File: python/analyzer.py
"""
analyzer.py — TaskAnalyzer: computes metrics, detects patterns, scores tasks
Showcases: classes, methods, type hints, list comprehensions, generators, property, staticmethod
"""

from __future__ import annotations

import math
import statistics
from collections import Counter, defaultdict
from typing import Dict, Generator, List, Optional, Tuple

from models import Priority, Status, Task, TaskCollection, TaskFilter


# ── Metric result types ───────────────────────────────────────────────────────

class MetricSnapshot:
    """Immutable snapshot of TaskCollection metrics at a point in time."""

    def __init__(
        self,
        total: int,
        by_status: Dict[str, int],
        by_priority: Dict[str, int],
        completion_rate: float,
        overdue_count: int,
        avg_age_days: float,
        priority_debt_score: float,
    ) -> None:
        self.total = total
        self.by_status = by_status
        self.by_priority = by_priority
        self.completion_rate = completion_rate
        self.overdue_count = overdue_count
        self.avg_age_days = avg_age_days
        self.priority_debt_score = priority_debt_score

    def __repr__(self) -> str:
        return (
            f"MetricSnapshot(total={self.total}, "
            f"completion={self.completion_rate:.1%}, "
            f"overdue={self.overdue_count})"
        )

    def as_dict(self) -> dict:
        return {
            "total": self.total,
            "by_status": self.by_status,
            "by_priority": self.by_priority,
            "completion_rate": self.completion_rate,
            "overdue_count": self.overdue_count,
            "avg_age_days": self.avg_age_days,
            "priority_debt_score": self.priority_debt_score,
        }


class HealthReport:
    """Human-readable health assessment of a task collection."""

    THRESHOLDS = {
        "critical_overdue": 10,
        "high_backlog": 50,
        "low_completion": 0.3,
    }

    def __init__(self, snapshot: MetricSnapshot) -> None:
        self.snapshot = snapshot
        self.warnings: List[str] = []
        self.recommendations: List[str] = []
        self._evaluate()

    def _evaluate(self) -> None:
        s = self.snapshot
        if s.overdue_count >= self.THRESHOLDS["critical_overdue"]:
            self.warnings.append(f"{s.overdue_count} overdue tasks — critical backlog!")
            self.recommendations.append("Triage overdue tasks immediately.")
        if s.completion_rate < self.THRESHOLDS["low_completion"]:
            self.warnings.append(f"Low completion rate: {s.completion_rate:.0%}")
            self.recommendations.append("Break large tasks into smaller subtasks.")
        if s.total >= self.THRESHOLDS["high_backlog"]:
            self.warnings.append(f"Large backlog: {s.total} total tasks")
            self.recommendations.append("Consider archiving or cancelling stale tasks.")

    @property
    def is_healthy(self) -> bool:
        return len(self.warnings) == 0

    def summary(self) -> str:
        status = "✅ Healthy" if self.is_healthy else "⚠️ Issues detected"
        lines = [status] + [f"  - {w}" for w in self.warnings]
        return "\n".join(lines)


# ── Core analyzer ─────────────────────────────────────────────────────────────

class TaskAnalyzer:
    """
    Computes metrics, ranks tasks, and detects quality patterns in a TaskCollection.
    """

    def __init__(self, collection: TaskCollection) -> None:
        self._collection = collection

    # ── Metrics ───────────────────────────────────────────────────────────

    def snapshot(self) -> MetricSnapshot:
        tasks = self._collection.all()
        if not tasks:
            return MetricSnapshot(0, {}, {}, 0.0, 0, 0.0, 0.0)

        by_status = Counter(t.status.value for t in tasks)
        by_priority = Counter(t.priority.name for t in tasks)
        done = by_status.get(Status.DONE.value, 0)
        completion_rate = done / len(tasks)
        overdue = sum(1 for t in tasks if t.is_overdue)
        ages = [t.age_days for t in tasks]
        avg_age = statistics.mean(ages) if ages else 0.0
        debt = self._priority_debt(tasks)

        return MetricSnapshot(
            total=len(tasks),
            by_status=dict(by_status),
            by_priority=dict(by_priority),
            completion_rate=completion_rate,
            overdue_count=overdue,
            avg_age_days=avg_age,
            priority_debt_score=debt,
        )

    def health_report(self) -> HealthReport:
        return HealthReport(self.snapshot())

    # ── Scoring ───────────────────────────────────────────────────────────

    def score(self, task: Task) -> float:
        """
        Composite urgency score: higher = should be done sooner.
        Factors: priority weight, age, overdue penalty, effort estimate.
        """
        base = task.priority.weight() * 10.0
        age_factor = math.log1p(task.age_days) * 2.0
        overdue_bonus = 20.0 if task.is_overdue else 0.0
        effort_penalty = task.estimate_effort() * 0.5
        return base + age_factor + overdue_bonus - effort_penalty

    def rank(self, top_n: Optional[int] = None) -> List[Tuple[Task, float]]:
        """Return tasks ranked by urgency score (descending)."""
        active = TaskFilter().with_status(*Status.active_states()).apply(self._collection.all())
        ranked = sorted(
            ((t, self.score(t)) for t in active),
            key=lambda pair: pair[1],
            reverse=True,
        )
        return ranked[:top_n] if top_n is not None else ranked

    # ── Pattern detection ─────────────────────────────────────────────────

    def find_blocked_chains(self) -> List[List[Task]]:
        """
        Detect tasks that are BLOCKED with subtasks also stuck.
        Returns chains (parent + blocked subtasks).
        """
        chains: List[List[Task]] = []
        for task in self._collection:
            if task.status == Status.BLOCKED:
                blocked_subs = [s for s in task._subtasks if s.status == Status.BLOCKED]
                if blocked_subs:
                    chains.append([task] + blocked_subs)
        return chains

    def stale_tasks(self, min_age_days: int = 30) -> List[Task]:
        """Return TODO tasks that haven't been updated in a long time."""
        return [
            t for t in self._collection
            if t.status == Status.TODO and t.age_days >= min_age_days
        ]

    def tag_distribution(self) -> Dict[str, int]:
        """Count tasks per tag."""
        dist: Dict[str, int] = defaultdict(int)
        for task in self._collection:
            for tag in task.tags:
                dist[tag.name] += 1
        return dict(dist)

    # ── Generators ────────────────────────────────────────────────────────

    def yield_critical(self) -> Generator[Task, None, None]:
        """Lazily yield CRITICAL priority tasks not yet done."""
        for task in self._collection:
            if task.priority == Priority.CRITICAL and not task.status.is_terminal():
                yield task

    # ── Private helpers ───────────────────────────────────────────────────

    @staticmethod
    def _priority_debt(tasks: List[Task]) -> float:
        """
        Priority debt = sum of (weight * age_days) for all active high+ tasks.
        High value means important work has been sitting for too long.
        """
        debt = 0.0
        for task in tasks:
            if task.priority >= Priority.HIGH and not task.status.is_terminal():
                debt += task.priority.weight() * math.log1p(task.age_days)
        return round(debt, 2)

    def _normalise_score(self, raw: float, max_score: float) -> float:
        if max_score == 0:
            return 0.0
        return min(raw / max_score, 1.0)

    # ── Never-called method (dead code demo) ──────────────────────────────

    def _legacy_report(self) -> str:
        """Old plaintext report — replaced by HealthReport. Dead code."""
        snap = self.snapshot()
        return f"Total: {snap.total}, Done: {snap.by_status.get('done', 0)}"

File: go/analyzer.go
// analyzer.go — Metrics, health, and pattern detection
// Showcases: closures, higher-order functions, error wrapping, struct embedding
package taskmanager

import (
	"errors"
	"fmt"
	"math"
)

// ── Errors ────────────────────────────────────────────────────────────────────

var (
	ErrEmptyCollection = errors.New("collection contains no tasks")
	ErrAnalysisFailed  = errors.New("analysis failed")
)

// ── MetricSnapshot ────────────────────────────────────────────────────────────

type MetricSnapshot struct {
	Total             int
	ByStatus          map[Status]int
	ByPriority        map[Priority]int
	CompletionRate    float64
	OverdueCount      int
	AvgAgeDays        float64
	PriorityDebtScore float64
}

func (s MetricSnapshot) String() string {
	return fmt.Sprintf(
		"Metrics{total=%d, done=%.0f%%, overdue=%d, debt=%.2f}",
		s.Total, s.CompletionRate*100, s.OverdueCount, s.PriorityDebtScore,
	)
}

// ── HealthReport ──────────────────────────────────────────────────────────────

type HealthReport struct {
	Snapshot        MetricSnapshot
	Warnings        []string
	Recommendations []string
}

func (h HealthReport) IsHealthy() bool { return len(h.Warnings) == 0 }

func (h HealthReport) Summary() string {
	if h.IsHealthy() {
		return "✅ Healthy"
	}
	msg := "⚠️ Issues:\n"
	for _, w := range h.Warnings {
		msg += "  - " + w + "\n"
	}
	return msg
}

// ── Analyzer ──────────────────────────────────────────────────────────────────

type Analyzer struct {
	manager *TaskManager
	// Pluggable scoring function — closures let callers customise scoring.
	ScoreFn func(t *Task) float64
}

func NewAnalyzer(manager *TaskManager) *Analyzer {
	defaultScore := func(t *Task) float64 { return t.Score() }
	return &Analyzer{manager: manager, ScoreFn: defaultScore}
}

func (a *Analyzer) Snapshot() (MetricSnapshot, error) {
	tasks := a.manager.All()
	if len(tasks) == 0 {
		return MetricSnapshot{}, ErrEmptyCollection
	}

	byStatus := make(map[Status]int)
	byPriority := make(map[Priority]int)
	totalAge := 0
	overdueCount := 0
	doneCount := 0

	for _, t := range tasks {
		byStatus[t.Status]++
		byPriority[t.Priority]++
		totalAge += t.AgeDays()
		if t.IsOverdue() {
			overdueCount++
		}
		if t.Status == StatusDone {
			doneCount++
		}
	}

	n := len(tasks)
	snapshot := MetricSnapshot{
		Total:             n,
		ByStatus:          byStatus,
		ByPriority:        byPriority,
		CompletionRate:    float64(doneCount) / float64(n),
		OverdueCount:      overdueCount,
		AvgAgeDays:        float64(totalAge) / float64(n),
		PriorityDebtScore: a.priorityDebt(tasks),
	}
	return snapshot, nil
}

func (a *Analyzer) HealthReport() (HealthReport, error) {
	snap, err := a.Snapshot()
	if err != nil {
		return HealthReport{}, fmt.Errorf("%w: %v", ErrAnalysisFailed, err)
	}
	report := HealthReport{Snapshot: snap}
	a.populateWarnings(&report)
	return report, nil
}

func (a *Analyzer) Rank(topN int) []*Task {
	return a.manager.RankByScore(topN)
}

// DetectPatterns returns named groups of tasks matching various antipatterns.
func (a *Analyzer) DetectPatterns() map[string][]*Task {
	// Using closures to define each pattern inline
	patterns := map[string]func(*Task) bool{
		"stale_todo": func(t *Task) bool {
			return t.Status == StatusTodo && t.AgeDays() > 30
		},
		"high_priority_blocked": func(t *Task) bool {
			return t.Status == StatusBlocked && t.Priority.IsUrgent()
		},
		"overdue_critical": func(t *Task) bool {
			return t.IsOverdue() && t.Priority == PriorityCritical
		},
		"zero_progress_subtasks": func(t *Task) bool {
			return len(t.Subtasks) > 0 && t.CompletionRate() == 0 && t.AgeDays() > 7
		},
	}

	result := make(map[string][]*Task, len(patterns))
	for name, pred := range patterns {
		result[name] = a.manager.Filter(pred)
	}
	return result
}

// WithCustomScorer returns a new Analyzer using a custom scoring closure.
func (a *Analyzer) WithCustomScorer(fn func(*Task) float64) *Analyzer {
	return &Analyzer{manager: a.manager, ScoreFn: fn}
}

// ── Private ───────────────────────────────────────────────────────────────────

func (a *Analyzer) priorityDebt(tasks []*Task) float64 {
	debt := 0.0
	for _, t := range tasks {
		if t.Priority.IsUrgent() && !t.Status.IsTerminal() {
			debt += float64(t.Priority.Weight()) * math.Log1p(float64(t.AgeDays()))
		}
	}
	return math.Round(debt*100) / 100
}

func (a *Analyzer) populateWarnings(report *HealthReport) {
	s := report.Snapshot
	if s.OverdueCount >= 10 {
		report.Warnings = append(report.Warnings, fmt.Sprintf("%d overdue tasks", s.OverdueCount))
		report.Recommendations = append(report.Recommendations, "Triage overdue tasks immediately")
	}
	if s.CompletionRate < 0.3 {
		report.Warnings = append(report.Warnings, fmt.Sprintf("Low completion: %.0f%%", s.CompletionRate*100))
		report.Recommendations = append(report.Recommendations, "Break large tasks into subtasks")
	}
	if s.PriorityDebtScore > 100 {
		report.Warnings = append(report.Warnings, fmt.Sprintf("High priority debt: %.1f", s.PriorityDebtScore))
	}
}

// ── Dead code ─────────────────────────────────────────────────────────────────

// legacyAverageScore was replaced by Rank(). Never called.
func (a *Analyzer) legacyAverageScore() float64 {
	tasks := a.manager.All()
	if len(tasks) == 0 {
		return 0
	}
	total := 0.0
	for _, t := range tasks {
		total += a.ScoreFn(t)
	}
	return total / float64(len(tasks))
}
